# Snakefile for general RNA-Seq analysis

import os
import pandas as pd
import requests
import urllib3
from pathlib2 import Path
from snakemake.utils import min_version
from snakemake.logging import logger

# Necessary to handle HTTP remote to download reference genome
requests.packages.urllib3.util.ssl_.DEFAULT_CIPHERS = 'ALL:@SECLEVEL=1'

# Minimal version of snakemake
min_version('6.8.0')

# Load Snakemake config 
configfile: "config/config.yaml"

# ---------------------------- Read config file ----------------------------- #

RESULTS_DIR = Path(config['RESULTS_DIR'])
FIGURE_DIR = Path(config['FIGURE_DIR'])
QC_DIR = Path(config['QC_DIR'])

EXTERNAL_DATA_DIR = Path(config['EXTERNAL_DATA_DIR'])
FASTQ_DIR = Path(config['FASTQ_DIR'])
GENOME_INDEX_DIR = Path(config['GENOME_INDEX_DIR'])

LOG_DIR = Path(config['LOG_DIR'])
TEMP_DIR = Path(config['TEMP_DIR'])
REPORT_DIR = Path(config['REPORT_DIR'])

SUBSET_COLUMN = config['SUBSET_COLUMN']

# --------------------------- Read sample manifest ---------------------------- #
SAMPLE_MANIFEST = pd.read_table(Path(config['SAMPLE_MANIFEST']))
SAMPLE_MANIFEST['Bio_rep'] = SAMPLE_MANIFEST.Bio_rep.astype(str)

# Specify the sample info combinations to use as label per sample
SAMPLES = SAMPLE_MANIFEST.set_index(['Sample_type',
                                     'Treatment',
                                     'Bio_rep'])


# ------------------------------ Dictitonaries ------------------------------- #

# GENCODE releases dictionary (release -> genome name # UCSC genome equivalent)
GENCODE_VERSIONS = {
    'M25': 'GRCm38', # mm10
    'M32': 'GRCm39', # mm39
    '19': 'GRCh37', # h19
    '38': 'GRCh38', # h38 
    # 'M1': 'NCBI37' # mm9 uses different handle
}

UCSC_VERSIONS = {
    'M25': 'mm10', # GRCm38
    'M32': 'mm39', # GRCm39
    '19': 'h19', # GRCh37
    '38': 'h38', # GRCh38 
    # 'M1': 'mm9' # NCBI37 uses different handle
}

# Output files dictionary
TARGETS = {# -- Raw procesing
            'raw_seq': [], 
            'merged_fq': [],
            'trimmed': [],
            'hisat2': [],
            'samtools': [],
            'samtools_index': [],
            'salmon': [],
            # -- Downstream analysis
            'te_count': [],
            'unaligned_reads': [],
            # -- All samples 
            'feature_counts': [],
            'te_mtx': [],
            'tximeta' : [],
            'ensembl' : [],
            'deseq2' : [],
            'counts' : [],
            # -- Reports
            'de_rmd': []
            }
    
# as array for different genomes?
GENCODE_RELEASE = pd.unique(SAMPLES['Target_genome'])[0]
GENOME = GENCODE_VERSIONS[GENCODE_RELEASE]
UCSC_GENOME = UCSC_VERSIONS[GENCODE_RELEASE]

for sample_type, treatment, bio_rep in SAMPLES.index:
    identifier = SAMPLES.loc[(sample_type, treatment, bio_rep), 'Identifier']
    handle = SAMPLES.loc[(sample_type, treatment, bio_rep), 'Fastq_handle']
    # Selects samples with specified value on a given column in the sample manifest (USE CONFIG FILE TO SELECT)
    if SUBSET_COLUMN:
        if SAMPLES.loc[(sample_type, treatment, bio_rep), SUBSET_COLUMN] == config['SUBSET_VALUE']:
            # --------- Raw procesing
            TARGETS['raw_seq'].append(str(FASTQ_DIR / 'raw'/ f'{identifier}_S{handle}_L002_R2_001.fastq.gz'))
            TARGETS['merged_fq'].append(str(FASTQ_DIR / 'merged'/ f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}_R2.fq.gz'))
            TARGETS['trimmed'].append(str(RESULTS_DIR / 'fastq' / 'trimmed' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}_R1.fq.gz_trimming_report.txt'))
            TARGETS['hisat2'].append(str(RESULTS_DIR / 'bam_files' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}.hisat2_align.bam'))
            TARGETS['samtools'].append(str(RESULTS_DIR / 'bam_files' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}.hisat2_align.sorted.filtered.bam.bai'))
            TARGETS['samtools_index'].append(str(RESULTS_DIR / 'bam_files' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}.hisat2_align.sorted.filtered.bam.bai'))
            TARGETS['salmon'].append(str(RESULTS_DIR / 'read_counts' / 'salmon' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / 'quant.sf'))
            # --------- Downstream analysis
            TARGETS['te_count'].append(str(RESULTS_DIR  / 'read_counts' / 'te_count' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}_TEcounts.cntTable'))
            TARGETS['unaligned_reads'].append(str(QC_DIR / 'unaligned_reads' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}.blast.unaligned_sequences.tsv'))
    else:
        # --------- Raw procesing
        TARGETS['raw_seq'].append(str(FASTQ_DIR / 'raw'/ f'{identifier}_S{handle}_L002_R2_001.fastq.gz'))
        TARGETS['merged_fq'].append(str(FASTQ_DIR / 'merged'/ f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}_R2.fq.gz'))
        TARGETS['trimmed'].append(str(RESULTS_DIR / 'fastq' / 'trimmed' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}_R1.fq.gz_trimming_report.txt'))
        TARGETS['hisat2'].append(str(RESULTS_DIR / 'bam_files' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}.hisat2_align.bam'))
        TARGETS['samtools'].append(str(RESULTS_DIR / 'bam_files' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}.hisat2_align.sorted.filtered.bam'))
        TARGETS['samtools_index'].append(str(RESULTS_DIR / 'bam_files' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}.hisat2_align.sorted.filtered.bam.bai'))
        TARGETS['salmon'].append(str(RESULTS_DIR / 'read_counts' / 'salmon' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / 'quant.sf'))
        # --------- Downstream analysis
        TARGETS['te_count'].append(str(RESULTS_DIR  / 'read_counts' / 'te_count' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}_TEcounts.cntTable'))
        TARGETS['unaligned_reads'].append(str(QC_DIR / 'unaligned_reads' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}' / f'{sample_type}_{treatment}_Bio-rep_{bio_rep}.blast.unaligned_sequences.tsv'))

# --------- All samples 
TARGETS['feature_counts'].append(str(RESULTS_DIR / 'read_counts' / 'featureCounts' / 'featureCounts_allSamples_genecounts.tsv'))
TARGETS['te_mtx'].append(str(RESULTS_DIR / 'read_counts' / 'te_count' / 'allSamples' / 'TEcount_allSamples_geneTEcounts.tsv'))
TARGETS['tximeta'].append(str(RESULTS_DIR / 'read_counts' / 'tximeta' / 'GeneSE_counts.tsv'))
TARGETS['ensembl'].append(str(EXTERNAL_DATA_DIR / 'genesets' / f'{GENOME}_ensembl_geneset.tsv'))
TARGETS['deseq2'].append(str(RESULTS_DIR / 'downstream' / 'differential_expr' / 'dds.Rds'))
TARGETS['deseq2'].append(str(RESULTS_DIR / 'downstream' / 'differential_expr' / 'DEGs' / 'DEGs_frequency.tsv'))
# --------- Reports
TARGETS['de_rmd'].append(str( REPORT_DIR / 'DE_analysis.html'))



if config['ALIGNER'] == "HISAT2":
    TARGETS['counts'].append(TARGETS['feature_counts'])
elif config['ALIGNER'] == "SALMON":
    TARGETS['counts'].append(TARGETS['tximeta'])
elif config['ALIGNER'] == "SALMON-TECOUNT":
    TARGETS['counts'].append(TARGETS['te_mtx'])
# elif config['ALIGNER'] == "BOTH":
#     TARGETS['counts'].append(TARGETS['feature_counts'], TARGETS['salmon'])

# ------------------------------- Import rules -------------------------------- #

# --------- Raw procesing 
include: 'rules/external_data/merge_lanes.smk'
include: 'rules/alignment/trim_galore.smk'
include: 'rules/external_data/get_ref_sequences.smk'
include: 'rules/alignment/hisat2.smk'
include: 'rules/alignment/samtools.smk'
include: 'rules/alignment/salmon.smk'

# --------- Downstream analysis
include: 'rules/alignment/samtools_strand.smk'
include: 'rules/quality_control/unaligned_read_blast.smk'
include: 'rules/read_counts/counts_parsing.smk'
include: 'rules/external_data/get_ensembl_geneset.smk'
include: 'rules/downstream/deseq2.smk'
include: 'rules/downstream/rep_elements.smk'

# --------- Report
include: 'rules/report/generate_Rmd.smk'

rule multiqc_report:
    input:
        TARGETS['counts']
    output:
        base = directory(str(REPORT_DIR / 'multiqc'))
    params:
        # LOG_DIR = Path(config['LOG_DIR']),
        # TEMP_DIR = Path(config['TEMP_DIR']),
        # REPORT_DIR = Path(config['REPORT_DIR'])
    log:
        str(LOG_DIR / 'multiqc' / 'multiqc.log')
    threads: 1
    conda:
        'envs/quality_control/multiqc.yaml'
    shell:
        """
        multiqc --force --interactive \
        -o {output.base} -n multiqc_report.html \
        {REPORT_DIR} {RESULTS_DIR} {LOG_DIR} &> {log}
        """

rule all:
    input:
        expand('resources/external/gencode_{release}/{genome}_genome.fa.gz', release=GENCODE_RELEASE, genome=GENOME),
        TARGETS['raw_seq'],
        TARGETS['merged_fq'],
        TARGETS['trimmed'],
        TARGETS['counts'],
        TARGETS['samtools_index'],
        TARGETS['te_count'],
        str(REPORT_DIR / 'multiqc'),
        # TARGETS['unaligned_reads']
        # TARGETS['tximeta'],
        # TARGETS['ensembl'],
        # TARGETS['deseq2']
        # TARGETS['rmd']

    default_target: True
